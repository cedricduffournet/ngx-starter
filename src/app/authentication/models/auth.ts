export interface AuthToken {
  access_token: string;
  expires_in: number;
  token_type: string;
  refresh_token: string;
}

export interface Credentials {
  login: string;
  password: string;
}

export interface ResetPassword {
  plainPassword: {
    first: string;
    second: string;
  };
}

export interface Register extends ResetPassword {
  civility: number;
  firstname: string;
  lastname: string;
  email: string;
  confirmationUrl: string;
}

export interface SendResetPassword {
  username: string;
}
