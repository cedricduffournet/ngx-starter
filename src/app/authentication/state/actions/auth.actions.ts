import { createAction } from '@ngrx/store';

export const loadLoggedUser = createAction('[Auth] Load logged user');
export const loadToken = createAction('[Auth] load token');
export const logout = createAction('[Auth] log out');
export const showLogoutModal = createAction(
  '[Auth] show log out modal confirm'
);
export const reloadUser = createAction('[Auth] Reload user');
