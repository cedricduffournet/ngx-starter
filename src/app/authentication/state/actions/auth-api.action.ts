import { props, createAction } from '@ngrx/store';

import { User } from '@app/user/models/User';
import { AuthToken } from '@app/authentication/models/auth';

export const loginSuccess = createAction(
  '[Auth/API] Login success',
  props<{ authToken: AuthToken }>()
);

export const loginFailure = createAction(
  '[Auth/API] Login failure',
  props<{ error: any }>()
);

export const registerSuccess = createAction(
  '[Auth/API] Register success',
  props<{ userRegistered: User }>()
);

export const registerFailure = createAction(
  '[Auth/API] Register failure',
  props<{ error: any }>()
);

export const loadLoggedUserSuccess = createAction(
  '[Auth/API] Load logged user success',
  props<{ user: User }>()
);

export const loadLoggedUserFailure = createAction(
  '[Auth/API] Load logged user failure',
  props<{ error: any }>()
);

export const loadTokenSuccess = createAction(
  '[Auth/API] load token success',
  props<{ authToken: AuthToken }>()
);

export const loadTokenFailure = createAction(
  '[Auth/API] load token failure',
  props<{ error: string }>()
);

export const logoutSuccess = createAction('[Auth] log out success');
