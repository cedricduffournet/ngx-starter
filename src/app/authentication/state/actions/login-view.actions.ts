import { props, createAction } from '@ngrx/store';
import { Credentials } from '@app/authentication/models/auth';

export const login = createAction(
  '[Login View] Login',
  props<{ credentials: Credentials }>()
);

export const clear = createAction('[Login View] Clear');
