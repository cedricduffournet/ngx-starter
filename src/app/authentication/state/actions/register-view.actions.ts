import { props, createAction } from '@ngrx/store';
import { Register } from '@app/authentication/models/auth';

export const register = createAction(
  '[Register View] Register user',
  props<{ register: Register }>()
);

export const clear = createAction('[Register View] Clear register view store');

export const loadCivilities = createAction(
  '[Register View] Load civility collection'
);
