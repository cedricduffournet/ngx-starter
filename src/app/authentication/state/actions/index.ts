import * as AuthActions from './auth.actions';
import * as AuthApiActions from './auth-api.action';
import * as LoginViewActions from './login-view.actions';
import * as RegisterViewActions from './register-view.actions';

export { AuthActions, AuthApiActions, LoginViewActions, RegisterViewActions };
