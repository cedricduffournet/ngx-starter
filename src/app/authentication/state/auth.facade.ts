import { Injectable } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs';

import * as fromAuth from '@app/authentication/state/reducers';

@Injectable()
export class AuthFacade {
  loggedUser$ = this.store.pipe(select(fromAuth.getLoggedUser));

  constructor(private store: Store<fromAuth.State>) {}

  isAuthorized(roles: string[]): Observable<boolean> {
    return this.store.pipe(select(fromAuth.getAuthorized, { roles }));
  }
}
