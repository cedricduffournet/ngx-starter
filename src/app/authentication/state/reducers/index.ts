import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers
} from '@ngrx/store';

import { User } from '@app/user/models/User';
import * as fromRoot from '@app/core/state/reducers';
import * as fromAuth from '@app/authentication/state/reducers/auth.reducer';
import * as fromLoginView from '@app/authentication/state/reducers/login-view.reducer';
import * as fromRegisterView from '@app/authentication/state/reducers/register-view.reducer';

export interface AuthState {
  userConnected: fromAuth.State;
  loginView: fromLoginView.State;
  registerView: fromRegisterView.State;
}

export interface State extends fromRoot.State {
  auth: AuthState;
}

export function reducers(state: AuthState | undefined, action: Action) {
  return combineReducers({
    userConnected: fromAuth.reducer,
    loginView: fromLoginView.reducer,
    registerView: fromRegisterView.reducer
  })(state, action);
}

export const selectAuthState = createFeatureSelector<State, AuthState>('auth');

export const getUserConnectedState = createSelector(
  selectAuthState,
  auth => auth.userConnected
);
export const getLoggedUser = createSelector(
  getUserConnectedState,
  auth => auth.user
);
export const getIsLogged = createSelector(
  getLoggedUser,
  user => !!user
);

export const getAuthorized = createSelector(
  getLoggedUser,
  (user: User | null, props: any) => {
    if (!user) {
      return false;
    }
    return props.roles.some((element: string) => {
      return user.roles.some((element2: string) => {
        return element2 === element;
      });
    });
  }
);

export const selectLoginViewState = createSelector(
  selectAuthState,
  (state: AuthState) => state.loginView
);
export const getLoginViewError = createSelector(
  selectLoginViewState,
  fromLoginView.getError
);
export const getLoginViewProcessing = createSelector(
  selectLoginViewState,
  fromLoginView.getProcessing
);

export const selectRegisterViewState = createSelector(
  selectAuthState,
  (state: AuthState) => state.registerView
);
export const getRegisterViewProcessing = createSelector(
  selectRegisterViewState,
  fromRegisterView.getProcessing
);
export const getRegisterViewStatus = createSelector(
  selectRegisterViewState,
  fromRegisterView.getStatus
);
export const getRegisterViewErrorUserExist = createSelector(
  selectRegisterViewState,
  fromRegisterView.getErrorUserExist
);
export const getRegisterViewUserRegistered = createSelector(
  selectRegisterViewState,
  fromRegisterView.getUserRegistered
);
