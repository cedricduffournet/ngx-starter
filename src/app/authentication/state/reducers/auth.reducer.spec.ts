import { reducer } from '@app/authentication/state/reducers/auth.reducer';
import * as fromAuth from '@app/authentication/state/reducers/auth.reducer';
import { User } from '@app/user/models/User';
import { AuthApiActions } from '@app/authentication/state/actions';

describe('AuthReducer ', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {} as any;

      const result = reducer(undefined, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOG_OUT_SUCCESS', () => {
    it('should logout user', () => {
      const initialState = {
        user: {
          firstname: 'firstname'
        }
      } as fromAuth.State;

      const action = AuthApiActions.logoutSuccess();
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_LOGGED_USER_FAILURE', () => {
    it('should unset logged user on logged failure', () => {
      const initialState = {
        user: {
          firstname: 'firstname'
        }
      } as fromAuth.State;
      const action = AuthApiActions.loadLoggedUserFailure({ error: {} });
      const result = reducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_LOGGED_USER_SUCCESS', () => {
    const user = {
      firstname: 'firstname',
      lastname: 'lastname'
    } as User;

    it('should load user info', () => {
      const action = AuthApiActions.loadLoggedUserSuccess({ user });
      const result = reducer(fromAuth.INITIAL_STATE, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getUser', () => {
      it('should return user', () => {
        const initialState = {
          user: {
            firstname: 'firstname',
            lastname: 'firstname'
          }
        } as fromAuth.State;

        const result = fromAuth.getUser(initialState);

        expect(result).toMatchSnapshot();
      });
    });
  });
});
