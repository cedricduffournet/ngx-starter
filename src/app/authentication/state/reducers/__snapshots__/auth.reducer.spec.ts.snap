// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthReducer  LOAD_LOGGED_USER_FAILURE should unset logged user on logged failure 1`] = `
Object {
  "user": null,
}
`;

exports[`AuthReducer  LOAD_LOGGED_USER_SUCCESS should load user info 1`] = `
Object {
  "user": Object {
    "firstname": "firstname",
    "lastname": "lastname",
  },
}
`;

exports[`AuthReducer  LOG_OUT_SUCCESS should logout user 1`] = `
Object {
  "user": null,
}
`;

exports[`AuthReducer  Selectors getUser should return user 1`] = `
Object {
  "firstname": "firstname",
  "lastname": "firstname",
}
`;

exports[`AuthReducer  undefined action should return the default state 1`] = `
Object {
  "user": null,
}
`;
